smallrye-reactive-messaging-aws-sns/src/main/java/io/smallrye/reactive/messaging/aws/sns/SnsConnector.java

```diff
@@ -0,0 +1,199 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.reactivex.Flowable;
+import io.reactivex.Scheduler;
+import io.reactivex.processors.MulticastProcessor;
+import io.reactivex.schedulers.Schedulers;
+import io.vertx.core.Vertx;
+import software.amazon.awssdk.services.sns.SnsAsyncClient;
+import software.amazon.awssdk.services.sns.model.CreateTopicRequest;
+import software.amazon.awssdk.services.sns.model.CreateTopicResponse;
+import software.amazon.awssdk.services.sns.model.PublishRequest;
+import software.amazon.awssdk.services.sns.model.PublishResponse;
+
+/**
+ * Implement incoming/outcoming connection factories for SNS reactive messaging.
+ * 
+ * @author iabughosh
+ * @version 1.0.4
+ *
+ */
+@ApplicationScoped
+@Connector(SnsConnector.CONNECTOR_NAME)
+public class SnsConnector implements IncomingConnectorFactory, OutgoingConnectorFactory {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SnsConnector.class);
+    static final String CONNECTOR_NAME = "smallrye-aws-sns";
+
+    /*
+     * Injectable fields
+     */
+    @Inject
+    @ConfigProperty(name = "sns-app-public-url")
+    Optional<String> appURL;
+    /** If there is existing verx instance, inject it. */
+    @Inject
+    Instance<Vertx> instanceOfVertx;
+    /*
+     * Fields
+     */
+    /** Control wither close vertx manually or not */
+    private boolean internalVertxInstance = false;
+
+    private Vertx vertx;
+    private ExecutorService threadExecutor;
+    private Scheduler scheduler;
+
+    /**
+     * Method being invoked when CDI bean first created.
+     */
+    @Inject
+    public void initConnector() {
+        //Initialize vertx and threadExecutor
+        LOGGER.info("Initializing Connector");
+        if (instanceOfVertx != null && instanceOfVertx.isUnsatisfied()) {
+            internalVertxInstance = true;
+            this.vertx = Vertx.vertx();
+        } else if (instanceOfVertx != null) {
+            this.vertx = instanceOfVertx.get();
+        }
+        threadExecutor = Executors.newSingleThreadExecutor();
+        scheduler = Schedulers.single();
+    }
+
+    /**
+     * Its being invoked before bean being destroyed.
+     */
+    @PreDestroy
+    public void preDestroy() {
+        LOGGER.info("Destroying Connector");
+        if (internalVertxInstance) {
+            Optional.ofNullable(vertx).ifPresent(vertx -> vertx.close());
+        }
+        Optional.ofNullable(threadExecutor).ifPresent(exec -> exec.shutdown());
+        Optional.ofNullable(scheduler).ifPresent(sch -> sch.shutdown());
+    }
+
+    @Override
+    public SubscriberBuilder<? extends Message<?>, Void> getSubscriberBuilder(Config config) {
+
+        String topicName = getTopicName(config);
+        return ReactiveStreams.<Message<?>> builder().flatMapCompletionStage(message -> {
+            CompletionStage<Message<?>> cs = CompletableFuture.runAsync(() -> {
+                //send to sns
+                send(message, topicName);
```

Comment:
If `send` returns a `CompletableFuture` (see comment), you can directly use it in the `flatMapCompletionStage` method. It would avoid using the executor.